### PR TITLE
readlink NFSv3 procedure

### PR DESCRIPTION
**Pull request by ploki**
Only the last commit

### CODE REVIEW

File: lib/nfsc.js [hunk 1]
```diff
@@ -696,6 +696,25 @@ class V3 {
...
+      * to the server.  That is, whether created by the NFS
+      * version 3 protocol software from a client or created
+      * locally on the server, the data in a symbolic link is not
+      * interpreted when created, but is simply stored.
+      *
+      * @param {object} object
+      * @param {function} callback(err: null,
+      *                            path: string,
+      *                            obj_attrs: object);
+      * @param {function} callback(err: string,
```
Comment (LaurenSpiegel) [inline lib/nfsc.js]: two callback params?

File: src/node_nfsc_readlink3.cc [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ * @authors:
+ *    Guillaume Gimenez <ggim@scality.com>
+ */
+#include "node_nfsc.h"
+#include "node_nfsc_readlink3.h"
+#include "node_nfsc_errors3.h"
+#include "node_nfsc_fattr3.h"
+
+// (object, callback(err, obj_attr) )
```
Comment (LaurenSpiegel) [inline src/node_nfsc_readlink3.cc]: buffer?

File: src/node_nfsc_readlink3.cc [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        return;
+    }
+    if (res.status != NFS3_OK) {
+        NFSC_ASPRINTF(&error, "%s", nfs3_error(res.status));
+        return;
+    }
+    success = true;
+}
+
+void NFS::ReadLink3Worker::HandleOKCallback()
```
Comment (LaurenSpiegel) [inline src/node_nfsc_readlink3.cc]: can this be abstracted for both read and set?

File: tests/functional/010_nfscV3.js [hunk 4]
```diff
@@ -232,9 +232,19 @@ async.waterfall([
                             assert.strictEqual(err, null);
                             assert.strictEqual(typeof(object), 'object');
                             assert.strictEqual(obj_attrs.type, mnt.NF3LNK);
-                            done(next, null);
+                            done(next, null, object);
```
Comment (LaurenSpiegel) [inline tests/functional/010_nfscV3.js]: why add object?

File: tests/functional/010_nfscV3.js [hunk 5]
```diff
@@ -232,9 +232,19 @@ async.waterfall([
...
+    (symlink, next) =>
+        describeIt('should read the symbolic link', done => {
+            mnt.readlink(symlink, (err, path, obj_attrs) => {
+                assert.strictEqual(err, null);
+                assert.strictEqual(typeof(obj_attrs), 'object');
+                assert.strictEqual(obj_attrs.mode, 0o777);
+                assert.strictEqual(path, 'foo-bar-baz');
+                done(next, null);
+            });
+        }),
```
Comment (LaurenSpiegel) [inline tests/functional/010_nfscV3.js]: test trying to read symlink that doesn't exist?

File: tests/functional/010_nfscV3.js (hunk 4 above)
Comment (ploki) [inline tests/functional/010_nfscV3.js]: for the next test in the row

File: lib/nfsc.js (hunk 1 above)
Comment (ploki) [inline lib/nfsc.js]: in case of error, the NFS server provides the client with additional information

File: src/node_nfsc_readlink3.cc (hunk 2 above)
Comment (ploki) [inline src/node_nfsc_readlink3.cc]: filesystem object handle stored in a buffer?

File: src/node_nfsc_readlink3.cc (hunk 3 above)
Comment (ploki) [inline src/node_nfsc_readlink3.cc]: you mean between symlink and readlink?

File: tests/functional/010_nfscV3.js [hunk 6]
```diff
@@ -224,6 +224,27 @@ async.waterfall([
...
+            var name = 'bar_' + crypto.randomBytes(8).toString('hex');
+            mnt.symlink(root_fh, name, {mode: 0o777}, 'foo-bar-baz',
+                        (err, object, obj_attrs, dir_wcc) => {
+                            assert.strictEqual(err, null);
+                            assert.strictEqual(typeof(object), 'object');
+                            assert.strictEqual(obj_attrs.type, mnt.NF3LNK);
+                            done(next, null, object);
+                        });
+        }),
+    (symlink, next) =>
```
Comment (ploki) [inline tests/functional/010_nfscV3.js]: you were right @LaurenSpiegel, ```symlink``` should be ```object```

File: src/node_nfsc_readlink3.cc (hunk 2 above)
Comment (LaurenSpiegel) [inline src/node_nfsc_readlink3.cc]: 👍

File: src/node_nfsc_readlink3.cc (hunk 3 above)
Comment (ploki) [inline src/node_nfsc_readlink3.cc]: Yes, there common contracts and logic that can be implemented in a parent class. And it's not limited to readlink/symlink.

File: tests/functional/010_nfscV3.js (hunk 5 above)
Comment (ploki) [inline tests/functional/010_nfscV3.js]: just testing the nominal case here. More complete tests of this module will come with the server as features show up.

**Pull request merged**